mapstructure.go

```diff
@@ -225,7 +225,15 @@ func (d *Decoder) Decode(input interface{}) error {
 // Decodes an unknown data type into a specific reflection value.
 func (d *Decoder) decode(name string, input interface{}, outVal reflect.Value) error {
 	if input == nil {
-		// If the input is nil, then we don't set anything.
+		// If the data is nil, then we don't set anything, unless ZeroFields is set
+		// to true.
+		if d.config.ZeroFields {
+			outVal.Set(reflect.Zero(outVal.Type()))
+
+			if d.config.Metadata != nil && name != "" {
```

Comment:
I am wondering if we should do the same thing in https://github.com/mitchellh/mapstructure/pull/109/files#diff-80aa7a137566e62ca9976c661b2b1a65R244 as well since we we are technically using the key and setting it to its zero value.